src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/DocumentPullDiagnosticsEndpoint.cs

```diff
@@ -52,58 +57,71 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagno
+                (csharpDiagnostics?.Length ?? 0) +
+                (htmlDiagnostics?.Length ?? 0));
 
-        if (razorDiagnostics is not null)
-        {
-            // No extra work to do for Razor diagnostics
-            allDiagnostics.AddRange(razorDiagnostics);
-        }
+            if (razorDiagnostics is not null)
+            {
+                // No extra work to do for Razor diagnostics
+                allDiagnostics.AddRange(razorDiagnostics);
+            }
 
-        if (csharpDiagnostics is not null)
-        {
-            foreach (var report in csharpDiagnostics)
+            if (csharpDiagnostics is not null)
             {
-                if (report.Diagnostics is not null)
+                foreach (var report in csharpDiagnostics)
                 {
-                    var mappedDiagnostics = await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.CSharp, report.Diagnostics, documentContext, cancellationToken);
-                    report.Diagnostics = mappedDiagnostics;
-                }
+                    if (report.Diagnostics is not null)
+                    {
+                        var mappedDiagnostics = await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.CSharp, report.Diagnostics, documentContext, cancellationToken);
+                        report.Diagnostics = mappedDiagnostics;
+                    }
 
-                allDiagnostics.Add(report);
+                    allDiagnostics.Add(report);
+                }
             }
-        }
 
-        if (htmlDiagnostics is not null)
-        {
-            foreach (var report in htmlDiagnostics)
+            if (htmlDiagnostics is not null)
             {
-                if (report.Diagnostics is not null)
+                foreach (var report in htmlDiagnostics)
                 {
-                    var mappedDiagnostics = await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.Html, report.Diagnostics, documentContext, cancellationToken);
-                    report.Diagnostics = mappedDiagnostics;
-                }
+                    if (report.Diagnostics is not null)
+                    {
+                        var mappedDiagnostics = await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.Html, report.Diagnostics, documentContext, cancellationToken);
+                        report.Diagnostics = mappedDiagnostics;
+                    }
 
-                allDiagnostics.Add(report);
+                    allDiagnostics.Add(report);
+                }
             }
+
+            return allDiagnostics.ToArray();
         }
+    }
 
-        return allDiagnostics.ToArray();
+    private IDisposable? Track(string name, Guid guid)
+    {
+        return _telemetryReporter.BeginBlock(name, Severity.Normal, ImmutableDictionary.CreateRange(new KeyValuePair<string, object?>[]
+        {
+            new("eventscope.method", "textdocument/_vs_diagnostic"),
```

Comment:
We could do this at a lower level too. Perhaps have a marker interface that we add to handlers, like `ITelemetryReportingHandler`, and have our queue do the tracking for any endpoint we mark.